File: src/app/toy/gerador-horario/horario-regras.ts
import { PencilmarkSolver, Possib } from "@/lib/pencilmark";
import { WorkerResponseMsg } from "@/lib/workerRace";
import { HorarioSolverPencilmark } from "./horario-pencilmark";
import { GeneticAlgorithm } from "@/lib/genetic/ga";
import { HorarioGAProblem } from "./horario-genetic";

export const TODOS_OS_DIAS = ["dom", "seg", "ter", "qua", "qui", "sex", "sab"] as const;
export const QUANTOS_DIAS = TODOS_OS_DIAS.length;
export type HorarioDia = typeof TODOS_OS_DIAS[number];

export const DIAS_SEMANA_MAP: Record<HorarioDia, string> = {
  dom: "Domingo",
  seg: "Segunda",
  ter: "Terça",
  qua: "Quarta",
  qui: "Quinta",
  sex: "Sexta",
  sab: "Sábado",
};
export type HorariosPorDia = Partial<Record<HorarioDia, number[]>>;

export interface FormularioHorario {
    turmas: {
        nome: string;
        horarios: HorariosPorDia;
    }[];
    disciplinas: {
        nome: string;
        turma: string;
        aulas: number;
        agrupar: number;
        dividir: boolean;
    }[];
    disciplinas_unidas: {
        grupo: string;
        disciplinas: string[];
    }[];
    professores: {
        nome: string;
        disciplinas: string[];
        horarios: HorariosPorDia;
    }[];
}

export interface TurmaHorarioResult {
    turma: string;
    horario: Array<{
        dia: HorarioDia;
        tempos: (string | null)[];
    }>;
}

export class Horario {
    static diaSemanaMap: Record<HorarioDia, number> = {
        dom: 0,
        seg: 1,
        ter: 2,
        qua: 3,
        qui: 4,
        sex: 5,
        sab: 6,
    };

    dias: number[][];

    constructor(tempos: HorariosPorDia, nTempos: number) {
        this.dias = TODOS_OS_DIAS.map((dia) => {
            // Inicia com array de 0 (indisponível)
            const diaTempos = new Array(nTempos).fill(0); 
            if(tempos[dia]) for (const tempo of tempos[dia]!) {
                if (tempo >= 1 && tempo <= nTempos) {
                    // marca como disponível (1)
                    diaTempos[tempo - 1] = 1;
                }
            }
            return diaTempos;
        });
    }

    possui(dia: number, tempo: number): boolean {
        return this.dias[dia][tempo] !== 0;
    }

    getHorarioDia(dia: number): number[] {
        return this.dias[dia];
    }

    copiar(outro: Horario) {
        for (let dia = 0; dia < this.dias.length; dia++) {
            for (let tempo = 0; tempo < this.dias[dia].length; tempo++) {
                this.dias[dia][tempo] = outro.dias[dia][tempo];
            }
        }
    }

    // Obtém o tempo com o valor mais alto
    static getMaxTempo(tempos: HorariosPorDia): number {
        let maxTempo = 0;
        for (const dia in tempos) {
            const temposDia = tempos[dia as HorarioDia];
            if (!temposDia) continue;
            
            const maxTempoDia = Math.max(...temposDia);
            if (maxTempoDia > maxTempo) {
                maxTempo = maxTempoDia;
            }
        }
        return maxTempo;
    }
}

export class Turma {
    id: number;
    nome: string;
    horarios: Horario;

    constructor(dados: {
        nome: string;
        horarios: HorariosPorDia;
    }, id: number, nTempos: number) {
        this.id = id;
        this.nome = dados.nome;
        this.horarios = new Horario(dados.horarios, nTempos);
    }
}

export class Professor {
    id: number;
    nome: string;
    // 1 = disponível, 0 = ocupado
    horarios: Horario;

    //  -1 = não dá aula
    //   0 = não está definido
    // > 0 = Índice da disciplina
    matriz: Horario;

    constructor(dados: {
        nome: string;
        disciplinas: string[];
        horarios: HorariosPorDia;
    }, id: number, nTempos: number) {
        this.id = id;
        this.nome = dados.nome;
        this.horarios = new Horario(dados.horarios, nTempos);
        this.matriz = new Horario(dados.horarios, nTempos);
    }
}

export class Disciplina {
    id: number;
    
    aulas: number;
    agrupar: number;
    dividir: boolean;
    nome: string;

    turma: Turma;
    disciplinasUnidas: Disciplina[];
    professores: Professor[];

    // contador de aulas alocadas, para controle durante a geração do quadro
    contAulas: number;

    constructor(dados: {
        nome: string;
        turma: string;
        aulas: number;
        agrupar: number;
        dividir: boolean;
    }, id: number) {
        this.id = id;
        this.turma = {} as unknown as Turma;
        this.disciplinasUnidas = [];
        this.professores = [];
        this.aulas = dados.aulas;
        this.agrupar = dados.agrupar;
        this.dividir = dados.dividir;
        this.nome = dados.nome;

        this.contAulas = 0;
    }

    possuiProfessor(prof: Professor): boolean {
        return this.professores.includes(prof);
    }
}

export abstract class RegrasHorario {
    turmas: Turma[];
    disciplinas: Disciplina[];
    professores: Professor[];
    nTempos: number;

    constructor(formData: FormularioHorario) {        
        // Determina o número de tempos a partir dos dados das turmas e professores
        this.nTempos = 0;
        for (const turma of formData.turmas) {
            const turmaMax = Horario.getMaxTempo(turma.horarios);
            if (turmaMax > this.nTempos) {
                this.nTempos = turmaMax;
            }
        }

        // 1. Inicializar entidades (na ordem das dependências)
        this.turmas = [];
        this.disciplinas = [];
        this.professores = [];
        for(const turma of formData.turmas) {
            const turmaObj = new Turma(turma, this.turmas.length, this.nTempos);
            this.turmas.push(turmaObj);
        }

        for(const professor of formData.professores) {
            const profObj = new Professor(professor, this.professores.length, this.nTempos);
            this.professores.push(profObj);
        }

        for(const disciplina of formData.disciplinas) {
            const discObj = new Disciplina(disciplina, this.disciplinas.length);
            this.disciplinas.push(discObj);

            const turma = this.turmas.find((t) => t.nome === disciplina.turma);
            if (!turma) {
                throw new Error(`Turma "${disciplina.turma}" não encontrada para disciplina "${disciplina.nome}"`);
            }
            discObj.turma = turma;
        }

        for(const professor of formData.professores) {
            const profObj = this.professores.find((p) => p.nome === professor.nome)!;
            
            for(const nomeDisc of professor.disciplinas) {
                const discObj = this.disciplinas.find((d) => d.nome === nomeDisc);
                if (!discObj) {
                    throw new Error(`Disciplina "${nomeDisc}" não encontrada para professor "${professor.nome}"`);
                }

                discObj.professores.push(profObj);
            }
        }

        for(const formDu of formData.disciplinas_unidas) {
            for(const nomeDisc of formDu.disciplinas) {
                const discObj = this.disciplinas.find((d) => d.nome === nomeDisc);
                if (!discObj) {
                    throw new Error(`Disciplina "${nomeDisc}" não encontrada para disciplina unida "${formDu.grupo}"`);
                }

                for(const nomeDiscUnida of formDu.disciplinas) {
                    if (nomeDiscUnida === nomeDisc) continue;

                    const discUnida = this.disciplinas.find((d) => d.nome === nomeDiscUnida);
                    if (!discUnida) {
                        throw new Error(`Disciplina "${nomeDiscUnida}" não encontrada para disciplina unida "${formDu.grupo}"`);
                    }

                    discObj.disciplinasUnidas.push(discUnida);
                }
            }
        }
    }

    toQuadroIndex(idTurma: number, dia: number, tempo: number): number {
        return (idTurma * QUANTOS_DIAS * this.nTempos) + (dia * this.nTempos) + tempo;
    }

    fromQuadroIndex(index: number): { idTurma: number; dia: number; tempo: number } {
        const idTurma = Math.floor(Math.floor(index / this.nTempos) / QUANTOS_DIAS);
        const dia = Math.floor(index / this.nTempos) % QUANTOS_DIAS;
        const tempo = index % this.nTempos;

        return { idTurma: idTurma, dia, tempo };
    }

    getQuadroValor(quadro: number[], idTurma: number, dia: number, tempo: number): number {
        if (idTurma < 0 || idTurma >= this.turmas.length) {
            return -1;
        }
        if (dia < 0 || dia >= QUANTOS_DIAS) {
            return -1;
        }
        if (tempo < 0 || tempo >= this.nTempos) {
            return -1;
        }

        return quadro[this.toQuadroIndex(idTurma, dia, tempo)];
    }

    getQuadroVazio(): number[] {
        // Preencher quadro com 0 onde pode ter aulas e -1 onde não haverá aulas
        const quadro = Array.from(
            { length: this.turmas.length * QUANTOS_DIAS * this.nTempos }, 
            () => -1
        );
        for (const turma of this.turmas) {
            for (let dia = 0; dia < QUANTOS_DIAS; dia++) {
                for (let tempo = 0; tempo < this.nTempos; tempo++) {
                    if (turma.horarios.possui(dia, tempo)) {
                        quadro[this.toQuadroIndex(turma.id, dia, tempo)] = 0; // marca como possível
                    }
                }
            }
        }

        return quadro;
    }

    toStatusString(quadro: number[], maxLength: number = 64): string {
        let retStr = "";
        for(let turma = 0; turma < this.turmas.length; turma++) {
            retStr += `Turma ${this.turmas[turma].nome}:\n`;
            for(let dia = 0; dia < QUANTOS_DIAS; dia++) {
                for(let tempo = 0; tempo < this.nTempos; tempo++) {
                    const val = quadro[this.toQuadroIndex(turma, dia, tempo)];
                    const disciplina = val > 0 ? this.disciplinas[val - 1].nome : (val === 0 ? "???" : "---");
                    retStr += `${disciplina}\t`;
                }
                retStr += "\n";
            }
            retStr += "\n";
        }
        return retStr;
    }
}

File: rust-wasm/src/horario/horario_regras.rs
use std::{collections::HashMap};

// Tipo a ser serializado-deserializado para comunicação com o frontend
use serde::{Deserialize};

pub const QUANTOS_DIAS: usize = 7; // Domingo a Sábado
#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum DiaSemana {
    Domingo = 0,
    Segunda = 1,
    Terca   = 2,
    Quarta  = 3,
    Quinta  = 4,
    Sexta   = 5,
    Sabado  = 6,
}

impl DiaSemana {
    #[inline(always)]
    pub fn to_index(&self) -> usize {
        *self as usize
    }

    pub fn from_index(index: usize) -> Self {
        match index {
            0 => DiaSemana::Domingo,
            1 => DiaSemana::Segunda,
            2 => DiaSemana::Terca,
            3 => DiaSemana::Quarta,
            4 => DiaSemana::Quinta,
            5 => DiaSemana::Sexta,
            6 => DiaSemana::Sabado,
            _ => panic!("Índice inválido para DiaSemana: {}", index),            
        }
    }
}

impl From<&str> for DiaSemana {
    fn from(value: &str) -> Self {
        match value.to_lowercase().as_str() {
            "dom" => DiaSemana::Domingo,
            "seg" => DiaSemana::Segunda,
            "ter" => DiaSemana::Terca,
            "qua" => DiaSemana::Quarta,
            "qui" => DiaSemana::Quinta,
            "sex" => DiaSemana::Sexta,
            "sab" => DiaSemana::Sabado,
            _ => panic!("Valor inválido para DiaSemana: {}", value),
        }
    }
}

pub type HorariosPorDia = HashMap<String, Vec<i32>>;

#[derive(Deserialize, Debug, Clone)]
pub struct FomularioHorarioTurma {
    pub nome: String,
    pub horarios: HorariosPorDia,
}

#[derive(Deserialize, Debug, Clone)]
pub struct FomularioHorarioDisciplina {
    pub nome: String,
    pub turma: String,
    pub aulas: i32,
    pub agrupar: i32,
    pub dividir: bool,
}

#[derive(Deserialize, Debug, Clone)]
pub struct FomularioHorarioDisciplinaUnida {
    pub grupo: String,
    pub disciplinas: Vec<String>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct FomularioHorarioProfessor {
    pub nome: String,
    pub disciplinas: Vec<String>,
    pub horarios: HorariosPorDia,
}

#[derive(Deserialize, Debug, Clone)]
pub struct FomularioHorario {
    pub turmas: Vec<FomularioHorarioTurma>,
    pub disciplinas: Vec<FomularioHorarioDisciplina>,
    pub disciplinas_unidas: Vec<FomularioHorarioDisciplinaUnida>,
    pub professores: Vec<FomularioHorarioProfessor>,
}

pub struct Horario {
    // 2D: 7 dias x n tempos, armazenado como array linearizado
    pub slots: Vec<i32>
}

impl Horario {
    pub fn new(n_tempos: usize) -> Self {
        // na prática o máximo seria 18 tempos x 7 dias x 4 bytes por i32 = 504 bytes
        Self {
            slots: vec![0; QUANTOS_DIAS * n_tempos]
        }
    }

    pub fn from_formulario(tempos: HorariosPorDia, n_tempos: usize) -> Self {
        // Inicia com array de 0 (indisponível)
        let mut horario = Self::new(n_tempos);

        for (dia_str, tempos_vec) in tempos {
            let dia = DiaSemana::from(dia_str.as_str());
            for tempo in tempos_vec {
                if tempo >= 1 && tempo as usize <= n_tempos {
                    let idx = (tempo - 1) as usize;
                    //let slot_idx = dia.to_index() * n_tempos + idx;
                    let slot_idx = idx * QUANTOS_DIAS + dia.to_index();
                    horario.slots[slot_idx] = 1; // Marca como disponível
                }
            }
        }

        horario
    }

    /*pub fn get_horario_dia(&self, dia: DiaSemana) -> &[i32] {
        let start = dia.to_index() * self.n_tempos;
        &self.slots[start..(start + self.n_tempos)]
    }*/

    pub fn possui(&self, dia: DiaSemana, tempo: usize) -> bool {
        let idx = tempo * QUANTOS_DIAS + dia.to_index();
        self.slots[idx] == 1
    }
}

pub struct Turma {
    pub id: usize,
    pub nome: String,
    pub horarios: Horario,
}

pub struct Professor {
    pub id: usize,
    pub nome: String,
    // 1 = disponível, 0 = ocupado
    pub horarios: Horario,

    //  -1 = não dá aula
    //   0 = não está definido
    // > 0 = Índice da disciplina
    //_matriz:   Horario,
}

pub struct Disciplina {
    pub id: usize,
    pub nome: String,

    pub aulas: i32,
    pub agrupar: i32,
    pub dividir: bool,
    
    pub turma_id: usize,
    pub disciplinas_unidas: Vec<usize>,
    pub professores: Vec<usize>,

    // contador de aulas alocadas, para controle durante a geração do quadro
    //_aulas: i32,
}

pub type QuadroHorario = Vec<i32>;

pub struct RegrasHorario {
    pub turmas: Vec<Turma>,
    pub disciplinas: Vec<Disciplina>,    
    pub professores: Vec<Professor>,
    pub n_tempos: usize,
}

impl RegrasHorario {
    pub fn new(formulario: FomularioHorario) -> Self {
        // Determina o número de tempos a partir dos dados das turmas e professores
        let n_tempos = formulario.turmas.iter()
            .map(|t|
                t.horarios.values()
                    .map(|tempos_vec|
                        tempos_vec.iter().max().unwrap_or(&0)
                    )
                    .max().unwrap_or(&0)
            )
            .max().unwrap_or(&0).clone() as usize;
        assert!(n_tempos > 0, "Número de tempos deve ser maior que zero");

        // Converte os formulários em estruturas internas
        let mut turmas = Vec::new();
        let mut disciplinas = Vec::new();
        let mut professores = Vec::new();

        for (i, turma_form) in formulario.turmas.into_iter().enumerate() {
            turmas.push(Turma { 
                id: i, 
                nome: turma_form.nome, 
                horarios: Horario::from_formulario(turma_form.horarios, n_tempos)
            });
        }
        
        for (i, disc_form) in formulario.disciplinas.into_iter().enumerate() {
            let turma_id = turmas.iter().position(|t| t.nome == disc_form.turma)
                .expect(&format!("Turma '{}' não encontrada para disciplina '{}'", disc_form.turma, disc_form.nome));

            disciplinas.push(Disciplina {
                id: i,
                nome: disc_form.nome,
                aulas: disc_form.aulas,
                agrupar: disc_form.agrupar,
                dividir: disc_form.dividir,
                turma_id: turma_id,
                disciplinas_unidas: Vec::new(),
                professores: Vec::new(),
            });
        }

        
        for (i, prof_form) in formulario.professores.into_iter().enumerate() {
            for disc_nome in prof_form.disciplinas {
                let disc_id = disciplinas.iter().position(|d| d.nome == disc_nome)
                    .expect(&format!("Disciplina '{}' não encontrada para professor '{}'", disc_nome, prof_form.nome));
                disciplinas[disc_id].professores.push(i);
            }

            professores.push(Professor { 
                id: i, 
                nome: prof_form.nome, 
                horarios: Horario::from_formulario(prof_form.horarios, n_tempos)
            });
        }

        for disc_unida in formulario.disciplinas_unidas {
            for disc_nome in &disc_unida.disciplinas {
                let disc_id = disciplinas.iter().position(|d| disc_nome == &d.nome)
                    .expect(&format!("Disciplina '{}' não encontrada para grupo '{}'", disc_nome, disc_unida.grupo));

                for dics_unida_nome in &disc_unida.disciplinas {
                    if dics_unida_nome == disc_nome {
                        continue; // Não adiciona a si mesma
                    }

                    let disc_unida_id = disciplinas.iter().position(|d| dics_unida_nome == &d.nome)
                        .expect(&format!("Disciplina '{}' não encontrada para o grupos de disciplinas unidas '{}'", dics_unida_nome, disc_unida.grupo));
                    
                    disciplinas[disc_id].disciplinas_unidas.push(disc_unida_id);
                }
            }
        }

        Self {
            turmas,
            disciplinas,
            professores,
            n_tempos,
        }
    }

    pub fn to_quadro_index(&self, turma_id: usize, dia: DiaSemana, tempo: usize) -> usize {
        // Índice linearizado para acesso ao quadro
        (turma_id * QUANTOS_DIAS * self.n_tempos) + (dia.to_index() * self.n_tempos) + tempo
    }

    pub fn from_quadro_index(&self, index: usize) -> (usize, DiaSemana, usize) {
        let turma_id = (index / self.n_tempos) / QUANTOS_DIAS;
        let dia_idx = (index / self.n_tempos) % QUANTOS_DIAS;
        let tempo = index % self.n_tempos;

        (turma_id, DiaSemana::from_index(dia_idx), tempo)
    }

    pub fn get_quadro_vazio(&self) -> QuadroHorario {
        // -1 onde não terá nada
        let mut quadro = vec![-1; self.turmas.len() * QUANTOS_DIAS * self.n_tempos];

        for turma in &self.turmas {
            for dia_idx in 0..QUANTOS_DIAS {
                let dia = DiaSemana::from_index(dia_idx);
                for tempo in 0..self.n_tempos {
                    if turma.horarios.possui(dia, tempo) {
                        // Disponível para alocação
                        quadro[self.to_quadro_index(turma.id, dia, tempo)] = 0;
                    }
                }
            }
        }

        quadro
    }
}
